File: Shortest_Path/python/01_해킹.py
import heapq

def dijkstra(start):
    heap_data = []
    heapq.heappush(heap_data, (0, start))
    distance[start] = 0
    while heap_data:
        dist, new = heapq.heappop(heap_data)
        if distance[new] < dist:
            continue
        for i in adj[new]:
            cost = dist + i[1]
            if distance[i[0]] > cost:
                distance[i[0]] = cost
                heapq.heappush(heap_data, (cost, i[0]))

for _ in range(int(input())):
    n, d, c = map(int, input().split())
    adj = [[] for _ in range(n + 1)]
    distance = [1e9] * (n + 1)

    for _ in range(d):
        a, b, s = map(int, input().split())
        adj[b].append((a, s))

    dijkstra(c)

    count = 0
    max_distance = 0
    for i in distance:
        if i != 1e9:
            count += 1
            if max_distance < i:
                max_distance = i
    print(count, max_distance)

File: Shortest_Path/c++/01_해킹.cpp
// fast campus 강의
// https://www.acmicpc.net/problem/10282
// 2

#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int T, N, M, C;
int d[100001];

void dijkstra(vector<vector<pair<int, int>>> adj, int start) {
	priority_queue<pair<int, int>> pq;
	pq.push({ 0, start });
	d[start] = 0;
	while (!pq.empty()) {
		int cost = -pq.top().first;
		int x = pq.top().second;
		pq.pop();
		if (d[x] < cost)
			continue;
		for (int i = 0; i < adj[x].size(); i++) {
			int next = adj[x][i].first;
			int dist = cost + adj[x][i].second;
			if (dist < d[next]) {
				d[next] = dist;
				pq.push({ -dist, next });
			}
		}
	}
}

int main() {
	cin >> T;
	for (int t = 0; t < T; t++) {
		cin >> N >> M >> C;
		vector<vector<pair<int, int>>> arr(N + 1);
		fill(d, d + 100001, 1e9);
		for (int i = 0; i < M; i++) {
			int a, b, s;
			cin >> a >> b >> s;
			arr[b].push_back({ a, s });
		}
		dijkstra(arr, C);

		int cost = 0, max_dist = 0;
		for (int i = 1; i <= N; i++) {
			if (d[i] != 1e9) {
				cost++;
				max_dist = max(max_dist, d[i]);
			}
		}
		cout << cost << ' ' << max_dist << '\n';
	}
}